code/datums/actions/build_designator.dm

```diff
@@ -0,0 +1,243 @@
+	if(!construct_choice)
+		return
+	construct_type = construct_choice
+
+	owner.balloon_alert(owner, "[construct_type::name]")
+	cleanup_hologram()
+
+///Signal handler for the owner rotating
+/datum/action/ability/activable/build_designator/proc/on_owner_rotate(datum/source, dir, newdir)
+	SIGNAL_HANDLER
+	if(dir == newdir)
+		return
+	if(!hologram)
+		return
+	update_hologram(new_dir = newdir)
+
+//Wrapper for show_hologram
+/datum/action/ability/activable/build_designator/proc/show_hologram_call(mob/user, atom/target)
+	SIGNAL_HANDLER
+	INVOKE_ASYNC(src, PROC_REF(show_hologram), user, target)
+
+/// move or create a hologram on mousemove, and also start the cleanup timer and check turf validity
+/datum/action/ability/activable/build_designator/proc/show_hologram(mob/user, atom/target)
+	SIGNAL_HANDLER
+	var/turf/target_turf = get_turf(target)
+	if(!target_turf)
+		cleanup_hologram()
+		return
+
+	if(!hologram)
+		create_hologram(target_turf)
+	update_hologram(target_turf)
+	start_cleanup_timer()
+
+/// creates the hologram and quickly fades it in, step_size is increased to make movement smoother
+/datum/action/ability/activable/build_designator/proc/create_hologram(turf/target_turf)
+	var/atom/selected = construct_type
+	var/obj/effect/build_hologram/new_hologram = new(target_turf, selected, TRUE)
+	new_hologram.alpha = 0
+	new_hologram.layer = ABOVE_OBJ_LAYER
+	new_hologram.glide_size = 32
+	animate(new_hologram, 1 SECONDS, alpha = initial(new_hologram.alpha))
+	new_hologram.setDir(owner.dir)
+	hologram = new_hologram
+
+//Updates the hologram position and validity
+/datum/action/ability/activable/build_designator/proc/update_hologram(turf/target_turf = hologram.loc, new_dir = owner.dir)
+	if(!hologram)
+		return
+	if(hologram.loc != target_turf)
+		hologram.abstract_move(target_turf)
+
+	hologram.setDir(new_dir)
+	hologram.remove_filter("invalid_turf_filter")
+
+	if(!check_turf_validity(target_turf))
+		hologram.add_filter("invalid_turf_filter", 1, color_matrix_filter(rgb(233, 23, 23)))
+		return FALSE
+	return TRUE
+
+/// check if the turf is valid or not for the selected build type, and apply a matrix color if not
+/datum/action/ability/activable/build_designator/proc/check_turf_validity(turf/open/target_turf)
+	if(!isopenturf(target_turf))
+		return FALSE
+	if(!target_turf.allow_construction)
+		return FALSE
+	var/area/area = get_area(target_turf)
+	if(area.area_flags & NO_CONSTRUCTION)
+		return FALSE
+	if(isspaceturf(target_turf))
+		return FALSE
+	var/border_obj = construct_type::atom_flags & ON_BORDER
+	for(var/obj/object in target_turf)
+		if(!object.density)
+			continue
+		if(border_obj && (object.atom_flags & ON_BORDER) && !(object.dir & hologram.dir))
+			continue
+		return FALSE
+	return TRUE
+
+///Sets the cleanup timer
+/datum/action/ability/activable/build_designator/proc/start_cleanup_timer()
+	delete_timer()
+	cleanup_timer = addtimer(CALLBACK(src, PROC_REF(cleanup_hologram)), cleanup_time, TIMER_STOPPABLE)
```

Comment:
Can you use TIMER_UNIQUE here instead?
Not sure about making this spam recreate the timer but howbad can it be :clueless:

Reply:
this is just directly ripped off the xeno build pr you merged stinky